### PR TITLE
feat:Adding support for ZMPOP command

### LINKED ISSUE #3888: implement ZMPOP

### PR DESCRIPTION
**Pull request by guyzilla**
Fixes #3888 

Signed-off-by: Guy Flysher <guyzila@gmail.com>

### CODE REVIEW

File: src/server/zset_family_test.cc [hunk 1]
```diff
@@ -757,6 +789,73 @@ TEST_F(ZSetFamilyTest, ZAddBug148) {
...
+  EXPECT_THAT(resp, IntArg(2));
+
+  resp = Run({"zmpop", "1", "b", "MAX"});
+  EXPECT_THAT(resp, ContainsLabeledScoredArray("b", {{"b2", "2"}}));
+
+  // Count > 1.
+  resp = Run({"zadd", "c", "1", "c1", "2", "c2"});
+  EXPECT_THAT(resp, IntArg(2));
+
+  resp = Run({"zmpop", "1", "c", "MAX", "COUNT", "2"});
```
Comment (adiholden) [inline src/server/zset_family_test.cc:845]: Lets add a test were count > 1 and the key contains less elements

File: src/server/zset_family_test.cc [hunk 2]
```diff
@@ -757,6 +789,73 @@ TEST_F(ZSetFamilyTest, ZAddBug148) {
...
+  EXPECT_THAT(resp, ContainsLabeledScoredArray("c", {{"c1", "1"}, {"c2", "2"}}));
+
+  // First non empty set is not the first set.
+  resp = Run({"zadd", "d", "1", "d1"});
+  EXPECT_THAT(resp, IntArg(1));
+
+  resp = Run({"zmpop", "2", "empty", "d", "MAX"});
+  EXPECT_THAT(resp, ContainsLabeledScoredArray("d", {{"d1", "1"}}));
+
+  // TODO: Add tests that check that sets that appear after the first non empty set are not
```
Comment (adiholden) [inline src/server/zset_family_test.cc:855]: yes:) and also please add to the test scenarios above a check that the element was actually popped, you can use the zcard command for this

File: src/server/zset_family.cc [hunk 3]
```diff
@@ -2060,6 +2076,75 @@ void ZRemRangeGeneric(string_view key, const ZSetFamily::ZRangeSpec& range_spec,
   }
 }
 
+// Returns the key of the first non empty set found in the list of shard arguments.
+// Returns nullopt if none.
+std::optional<std::string> GetFirstNonEmptyKeyFound(EngineShard* shard, Transaction* t) {
```
Comment (adiholden) [inline src/server/zset_family.cc:2081]: I believe you can return string_view from GetFirstNonEmptyKeyFound
The ShardArgs are pointing to the command args that are allocated when parsing the command

File: src/server/zset_family.cc [hunk 4]
```diff
@@ -2355,12 +2440,70 @@ void ZSetFamily::ZInterCard(CmdArgList args, const CommandContext& cmd_cntx) {
...
+  // From the list of input keys, keeps the first (in the order of keys in the command) key found in
+  // the current shard.
+  absl::flat_hash_set<std::string> first_found_key_in_shard;
+  // We can have at most one result from each shard.
+  first_found_key_in_shard.reserve(std::min(shard_set->size(), num_keys));
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    std::optional<std::string> result = GetFirstNonEmptyKeyFound(shard, t);
+    if (result.has_value()) {
+      first_found_key_in_shard.insert(std::move(*result));
```
Comment (adiholden) [inline src/server/zset_family.cc:2461]: this callback is executed in different shards (threads) and can be done simultaneously, therefor we must do thread safe ops.
you can use std::vector size of shard_set->size()

File: src/server/zset_family.cc [hunk 5]
```diff
@@ -2355,12 +2440,70 @@ void ZSetFamily::ZInterCard(CmdArgList args, const CommandContext& cmd_cntx) {
...
+    }
+    return OpStatus::OK;
+  };
+
+  cmd_cntx.tx->Execute(std::move(cb), false /* possibly another hop */);
+
+  // Now that we have the first non empty key from each shard, find the first overall first key and
+  // pop elements from it.
+
+  auto* rb = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
```
Comment (adiholden) [inline src/server/zset_family.cc:2471]: why do you use both this and response_builder (line 2450)?

File: src/server/zset_family.cc [hunk 6]
```diff
@@ -2355,12 +2440,70 @@ void ZSetFamily::ZInterCard(CmdArgList args, const CommandContext& cmd_cntx) {
...
+  ArgRange arg_keys(args.subspan(1, num_keys));
+  // Find the first arg_key which exists in any shard and is not empty.
+  for (std::string_view key : arg_keys) {
+    if (first_found_key_in_shard.contains(key)) {
+      key_to_pop = key;
+      break;
+    }
+  }
+
+  if (key_to_pop.empty()) {
```
Comment (adiholden) [inline src/server/zset_family.cc:2483]: Empty string is also a valid key
you can run 
zadd "" 1 f1

File: src/server/zset_family.cc [hunk 7]
```diff
@@ -2060,6 +2076,75 @@ void ZRemRangeGeneric(string_view key, const ZSetFamily::ZRangeSpec& range_spec,
...
+  }
+
+  return std::optional<std::string>();
+}
+
+// Validates the ZMPop command arguments and extracts the values to the output params.
+// If the arguments are invalid sends the appropiate error to builder and returns false.
+bool ValidateZMPopCommand(CmdArgList args, uint32* num_keys, bool* is_max, int* pop_count,
+                          SinkReplyBuilder* builder) {
+  if (!SimpleAtoi(ArgS(args, 0), num_keys)) {
```
Comment (adiholden) [inline src/server/zset_family.cc:2104]: Please use CmdArgParser

File: src/server/zset_family.cc [hunk 8]
```diff
@@ -1821,31 +1820,48 @@ void ZBooleanOperation(CmdArgList args, string_view cmd, bool is_union, bool sto
...
+    if (!key_shard.has_value() || *key_shard == shard->shard_id()) {
+      result = std::move(OpPopCount(range_spec, t->GetOpArgs(shard), key));
+    }
+    return OpStatus::OK;
   };
 
-  OpResult<ScoredArray> result = tx->ScheduleSingleHopT(std::move(cb));
-  OutputScoredArrayResult(result, range_params, builder);
+  tx->Execute(std::move(cb), true);
+  tx->Conclude();
```
Comment (adiholden) [inline src/server/zset_family.cc:1849]: you dont need to call Conclude after Execute with with second arg = true

**Comment by adiholden** [comment]
@guyzilla This is a good PR. I gave you few comments. Thanks

File: src/server/zset_family_test.cc (hunk 1 above)
Comment (guyzilla) [reply src/server/zset_family_test.cc:845]: Done

File: src/server/zset_family_test.cc (hunk 2 above)
Comment (guyzilla) [reply src/server/zset_family_test.cc]: I used ZCARD to verify that sets are empty but ZRANGE for other cases since I wanted to verify that the correct elements are still there.

File: src/server/zset_family.cc (hunk 3 above)
Comment (guyzilla) [reply src/server/zset_family.cc]: Done

File: src/server/zset_family.cc (hunk 4 above)
Comment (guyzilla) [reply src/server/zset_family.cc]: Done

File: src/server/zset_family.cc (hunk 5 above)
Comment (guyzilla) [reply src/server/zset_family.cc]: Just an accident. Removed.

File: src/server/zset_family.cc (hunk 6 above)
Comment (guyzilla) [reply src/server/zset_family.cc]: Done

File: src/server/zset_family.cc (hunk 7 above)
Comment (guyzilla) [reply src/server/zset_family.cc]: Done

File: src/server/zset_family.cc (hunk 8 above)
Comment (guyzilla) [reply src/server/zset_family.cc]: Done

**Review by adiholden — APPROVED** [review]
